### Justin Ehly
#### HW6 - NLP
#### Answer explanations and discussion are after the code for each section.

1.	Evaluate text similarity of Amazon book search results by doing the following:<br>
a.	Do a book search on Amazon via the search box. Manually copy the full book title (including subtitle) of each of the top 24 books listed in the first two pages of search results. 


In [1]:
books = ['Before the Gospels: The Gospels of Thomas, Q, Signs, and The Passion: The Writings from Which the Gospels Sprang',
         'The Gnostic Gospels: The Gnostic Wisdom of Jesus including the Lost Apocryphal Gospel of Thomas, John and Mary Magdalene',
         'The Gospel Of Mary And Other Apocryphal Gospels: The Gospel Of Mary, Peter, Thomas, The Birth Of Mary And The Acts Of Pontius Pilate',
         'The Critical Edition of Q (Hermeneia)',
         'The Sayings Gospel Q in Greek and English with Parallels from the Gospels of Mark and Thomas (English and Greek Edition)',
         'Q Thomas Reader (English, Coptic and Coptic Edition)',
         'Thomas and the Gospels: The Case for Thomas\'s Familiarity with the Synoptics',
         'How the Words of Jesus Became Gospel: How the Sources of Q, Signs, Passion, and Thomas Compare to the New Testament',
         'The Gospel Q: The source gospel for Mark, Matthew, Luke and Thomas',
         'The Complete Gospel Parallels: Synopses of the Gospels Matthew, Mark, Luke, John, Thomas, Peter, Other Gospels and the Reconstructed Q Gospel by Arthur J. Dewey (15-Jan-2012) Paperback',
         'These Final Days Part 2: The Truth about How the Great Tribulation Begins, the Third Temple, and Jerusalem',
         'The Complete Books of Enoch (Annotated): 1 Enoch: The Ethiopian Book of Enoch, 2 Enoch: The Book of the Secrets of Enoch, and 3 Enoch: The Hebrew Book of Enoch (Collectors Edition)',
         'The Gnostic Gospels: Including the Gospel of Thomas, the Gospel of Mary Magdalene (Sacred Texts)',
         'The Lost Gospel Q: The Original Sayings of Jesus',
         'Lost Books of the Bible: The Great Rejected Texts',
         'Beyond Belief: The Secret Gospel of Thomas',
         'The Gnostic Bible: Revised and Expanded Edition',
         'The Complete Gospels, 4th Edition',
         'The Gospel of Thomas: A New Translation for Spiritual Seekers',
         'The Thomas Code: Solving the mystery of the Gospel of Thomas',
         'Jesus Through Middle Eastern Eyes: Cultural Studies in the Gospels',
         'The Lost Books of the Bible',
         'The Lost Gospel of Barabbas: Exodus (The Lost Gospel of Barabbas Trilogy Book 2)',
         'The Books of Enoch: Complete Edition: Including (1) The Ethiopian Book of Enoch, (2) The Slavonic Secrets and (3) The Hebrew Book of Enoch']



In [2]:
len(books)

24

b.	In Python, run one of the text-similarity measures covered in this course, e.g., cosine similarity. Compare each of the book titles, pairwise, to every other one. <br>

In [3]:
# build vectors and pairwise cosine similarities
import itertools
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

class homework():
    def __init__(self, books):
        self=self
        self.books = books

    def pairwise_comp(self):
        score_combinations = []
        for a,b in itertools.combinations(self.books, 2):
            vectorizer = TfidfVectorizer()
            vectors = vectorizer.fit_transform([a,b])
            score_combinations.append([round(cosine_similarity(vectors)[0][1],3),[a,b]])
        self.score_combinations = score_combinations

In [4]:
scores = homework(books)
scores.pairwise_comp()
print('Let\'s look at the first 3 similary scores and combinations of titles')
for i,x in scores.score_combinations[0:3]:
    print(f'Similarity score between \n{x[0]} \nand \n{x[1]} \nis {[i]}\n-------')


Let's look at the first 3 similary scores and combinations of titles
Similarity score between 
Before the Gospels: The Gospels of Thomas, Q, Signs, and The Passion: The Writings from Which the Gospels Sprang 
and 
The Gnostic Gospels: The Gnostic Wisdom of Jesus including the Lost Apocryphal Gospel of Thomas, John and Mary Magdalene 
is [0.478]
-------
Similarity score between 
Before the Gospels: The Gospels of Thomas, Q, Signs, and The Passion: The Writings from Which the Gospels Sprang 
and 
The Gospel Of Mary And Other Apocryphal Gospels: The Gospel Of Mary, Peter, Thomas, The Birth Of Mary And The Acts Of Pontius Pilate 
is [0.478]
-------
Similarity score between 
Before the Gospels: The Gospels of Thomas, Q, Signs, and The Passion: The Writings from Which the Gospels Sprang 
and 
The Critical Edition of Q (Hermeneia) 
is [0.272]
-------


---

c.	Which two titles are the most similar to each other? Which are the most dissimilar? Where do they rank, among the first 24 results?

In [5]:
score_comb = scores.score_combinations
score_comb = sorted(score_comb, key=lambda x: x[0], reverse=True)
print(f'The book titles with the most similarity has a score of {score_comb[0][0]} and were \n{score_comb[0][1][0]}\nand\n{score_comb[0][1][1]}')
print('---')
print(f'The book titles with the least similarity has a score of {score_comb[-1][0]} and were \n{score_comb[-1][1][0]}\nand\n{score_comb[-1][1][1]}')

The book titles with the most similarity has a score of 0.904 and were 
The Complete Books of Enoch (Annotated): 1 Enoch: The Ethiopian Book of Enoch, 2 Enoch: The Book of the Secrets of Enoch, and 3 Enoch: The Hebrew Book of Enoch (Collectors Edition)
and
The Books of Enoch: Complete Edition: Including (1) The Ethiopian Book of Enoch, (2) The Slavonic Secrets and (3) The Hebrew Book of Enoch
---
The book titles with the least similarity has a score of 0.0 and were 
Q Thomas Reader (English, Coptic and Coptic Edition)
and
The Lost Gospel of Barabbas: Exodus (The Lost Gospel of Barabbas Trilogy Book 2)


In [6]:
b = [score_comb[0][1][0],score_comb[0][1][1],score_comb[-1][1][0],score_comb[-1][1][1]]
import numpy as np
# find the rank of each title 
rank = [books.index(i) for i in b]
rank.sort()
print('This is where each title was ranked in the Amazon title list:')
for r in rank:
    print(f'{r+1}: {books[r]}')
    

This is where each title was ranked in the Amazon title list:
6: Q Thomas Reader (English, Coptic and Coptic Edition)
12: The Complete Books of Enoch (Annotated): 1 Enoch: The Ethiopian Book of Enoch, 2 Enoch: The Book of the Secrets of Enoch, and 3 Enoch: The Hebrew Book of Enoch (Collectors Edition)
23: The Lost Gospel of Barabbas: Exodus (The Lost Gospel of Barabbas Trilogy Book 2)
24: The Books of Enoch: Complete Edition: Including (1) The Ethiopian Book of Enoch, (2) The Slavonic Secrets and (3) The Hebrew Book of Enoch


---
2.	Now evaluate using a major search engine.<br>
a.	Enter one of the book titles from question 1a into Google, Bing, or Yahoo!. Copy the capsule of the first organic result and the 20th organic result. Take web results only (i.e., not video results), and skip sponsored results. 


In [7]:
queries = ['The Gospels of Thomas, Q, Signs, and The Passion ...https://www.amazon.com › Before-Gospels-Passion-Nar...Amazon.com: Before The Gospels: The Gospels of Thomas, Q, Signs, and The Passion Narrative: The       Writings from which the Gospels Sprang eBook : Lumpkin, ...Rating: 3.7 · 8 reviews',
          'The Four Gospels (Literature) - TV Tropeshttps://tvtropes.org › Literature A description of tropes appearing in Four Gospels. The first four books of the New Testament, chronicling the life of one Jesus Christ the son of David, ...']

---
b.	Run the same text similarity calculation that you used for question 1b on each of these capsules in comparison to the original query (book title). 

In [8]:
cos_scores = []
vectorizer = TfidfVectorizer()
for query in queries:
    vectors = vectorizer.fit_transform([books[0], query])
    cos_scores.append(cosine_similarity(vectors)[0][1])
print(f'The returned cosine similarity scores are {cos_scores[0]} and {cos_scores[1]}.')

The returned cosine similarity scores are 0.8579431731423922 and 0.4631339961861422.


In [11]:
sc = [s[0] for s in scores.score_combinations]
np.mean(sc)

0.2738188405797102

---
c.	Which one has the highest similarity measure? <br>
<font color='blue'>The first google search has the highest similary measure to the original book title with a score of 0.858 <br>
<br>
Insights:
For this project I searched for book titles relating to the ending of the async videos where the professor talks about Q and the Thomas gospel. I have no history of searching for any religous items on Amazon, so I thought it might make the assignming more interesting. The search results returned a lot of book titles that were appeared to be very similar to the original book title. In fact it returned one of the twice so I had to go back copy an additional one.<br>
After studying this unit it is interesting to see that Amazon appears to be recommending titles based on a combination of Cosine Similary. When I compared the google search results from result 1 to result 20, I notice that it is a higher similarity score than the mean of the Amazon search results.